### ISSUE TITLE
冻结 Gosmos Trainer Fork 与 Output Compatibility 契约

### ISSUE THREAD

**Issue by Italink** [Italink/Gosmos#42]
## Question

决定 `Gosmos Trainer Fork` 作为 owned `Trainer Backend` 的 seam 与 compatibility contract：submodule revision、`origin`/`upstream` 同步、capability/version fingerprint、entry point 与 argument surface 如何管理；哪些修改属于 fork implementation；怎样保证 current output 继续满足 standard Graphdeco PLY schema，并为未来 versioned PBR attributes 留出演进空间而不在本轮实现它们。

**Comment by Italink** [comment]
## Resolution

冻结 `Gosmos Trainer Fork` 的 ownership、invocation 与 output compatibility contract。这里的 ownership 表示 Gosmos 对维护与 compatibility seam 负责，不表示取得 Graphdeco 原始代码版权。

### Backend authority and provenance

- `Gosmos Trainer Fork` 是 Gosmos 唯一受支持、唯一可选择的 `Trainer Backend`。官方 `Graphdeco Trainer` 只作为 `upstream` source，不再进入 Editor settings、environment detection 或 Conversion backend selection。
- Gosmos repository 固定 Trainer top-level commit 与所有 nested submodule commits；当前 upstream baseline 为 Trainer `54c035f7834b564019656c3e3fcc3646292f727d`、rasterizer `9c5c2028f6fbee2be239bc4c9421ff894fe4fbe0`。
- `origin` 指向 `Italink/gaussian-splatting`，官方 repository 保持为 `upstream`。更新必须经过显式 maintenance branch、compatibility tests 与 benchmark，禁止跟随 moving branch 自动升级。
- 若修改 `diff-gaussian-rasterization`，必须先建立 Italink fork、把 nested submodule 指向该 fork，并保留官方 `dr_aa` source 为 upstream；未修改的 nested dependency 可以继续使用官方 pinned commit。
- 正常 `Conversion` 拒绝 dirty Trainer。Disposable prototype 可以显式运行 dirty tree，但结果必须标为 non-reproducible、禁止 cache，并记录 tree/patch fingerprint；该能力不进入普通 Conversion UI。

### Stable Trainer Contract

UE 不再直接把 `train.py` 与 Graphdeco flags 当作 interface。唯一 owned entry point 为：

```text
python -m gosmos_trainer capabilities --json
python -m gosmos_trainer train --request <trainer-request.json>
```

- `train.py`、Graphdeco argument groups 与具体 training loop 都属于 fork implementation。
- Versioned JSON request 只表达 Dataset manifest、output directory、training profile、seed 与经过筛选的 typed `Trainer Override`；禁止 free-form arguments，也不自动把新增 upstream flag 暴露给 UE。
- 现有 Graphdeco-style UI overrides 在迁移时逐项判断用户价值；保留项进入 request schema，由 fork 内部 adapter 映射到 implementation，未保留项回到 profile defaults。
- `trainer_interface_version` 使用 major compatibility；breaking request/event/meaning change 必须提升 major。Capability IDs 独立 version，Gosmos 按 required capability set 判断兼容，不再通过搜索 `train.py --help` 或 remote URL 判断。

### Capability and environment fingerprint

`capabilities --json` 必须声明：

- Trainer interface version、supported Dataset Profiles 与 `Output Profile`；
- coverage、inverse-Z、normal、density-control 等 versioned capability IDs；
- Trainer 与所有 modified nested submodule revisions、dirty state 和 native extension build IDs；
- Python、PyTorch、CUDA compatibility 与实际 module load paths。

每个 `Conversion Job` 的 `Trainer Fingerprint` 记录以上结果，再加 Python executable、GPU、request/config hash。用户仍可选择 Python executable，但 source/binary mismatch、错误 native extension build 或从其他 installation 加载 module 时 validation 必须失败。

### Fork implementation seams

- Native Dataset adapter、loss composition、density policy、capability document、request validation 与 process protocol 属于 top-level Python fork implementation。
- CUDA rasterizer 只承担 single-pass differentiable outputs/backward 与不可避免的 native plumbing；Dataset semantics、loss weights、training policy 与 output-profile policy不得下沉到 CUDA。
- Checkpoint 是 fork-internal artifact，只允许在相同 Trainer interface major、model-state schema、Trainer commit 与 relevant nested commits 下恢复。跨版本恢复必须由显式 migrator 完成；PLY 才是稳定 interchange representation。

### Process protocol

- stdout 使用 versioned JSON Lines，事件至少包括 `started`、`progress`、`checkpoint`、`output`、`completed` 与 `error`；Gosmos 不再解析 tqdm 或日志措辞。
- stderr 保留 human-readable diagnostics；failure 使用 stable categorized exit codes。
- cancellation 先发送 cooperative interrupt，在 bounded grace period 后才强制终止。
- Trainer 必须在 terminal event 与 process exit code 之间保持一致；缺失或矛盾按 protocol failure 处理。

### Output Profiles and verification

- 当前 output 固定为 `graphdeco-3dgs-ply/v1`：完整保留官方 property names、order、types、SH field count 与 binary little-endian layout。Geometry Supervision 只改变 training，不增加当前 PLY attributes。
- 未来 PBR 使用同一个 PLY container 内的 vertex attributes，定义为独立的 `gosmos-pbr-ply/v1` `Output Profile`；保留全部 Graphdeco base attributes，并使用 versioned `gosmos_pbr_*` attributes。PBR 的具体 attributes 与训练不在本 ticket 实现或冻结。
- Extended PBR PLY 不宣称是 exact standard Graphdeco schema；它必须能生成剥离 Gosmos attributes 的 `graphdeco-3dgs-ply/v1` compatibility projection。
- Trainer 先写 temporary output，再 atomic rename；同时写 output manifest，记录 Output Profile、Dataset/request hash、完整 `Trainer Fingerprint`、point count 与 PLY SHA-256。
- Gosmos 独立验证 manifest binding、完整 schema、property count/order/type、SH layout、finite values 与 declared hash，不能只验证文件存在或信任 Trainer 声明。

### License gate

当前 fork 继承的 [Gaussian-Splatting License](https://github.com/Italink/gaussian-splatting/blob/54c035f7834b564019656c3e3fcc3646292f727d/LICENSE.md) 将使用限制为 non-commercial research/evaluation。当前 map 与 prototype 采用该范围；任何 commercial distribution 或 production use 都必须先获得 Licensor 明确许可，或替换为 license-compatible implementation。维护 fork 不改变该限制。
